### PR TITLE
Integrate layerList to store and add breadcrumbs

### PR DESCRIPTION
**Pull request by junqiu-lei**
Signed-off-by: Junqiu Lei <junqiu@amazon.com>

### Description
* Integrate map layers to store
* Add saved map to route in the format of `app/maps-dashboards/{id}` 
* Add breadcrumbs to listing page, create and saved map page

### Issues Resolved
https://github.com/opensearch-project/dashboards-maps/issues/68

### Demo

https://user-images.githubusercontent.com/90288540/199593239-3b72cdc1-5b47-4093-8657-25aaa2567377.mov

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
Shall we update breadcrumb immediately with name once it is saved for first time?

**Comment by junqiu-lei** [comment]
> Shall we update breadcrumb immediately with name once it is saved for first time?

Makes sense to immediately update once saved. Will update.

File: maps_dashboards/common/index.ts [hunk 1]
```diff
@@ -18,7 +18,9 @@ export const MAP_INITIAL_STATE = {
 };
 
 export const APP_PATH = {
-  CREATE_MAP: '/create-map',
+  LANDING_PAGE_PATH: '/',
+  CREATE_MAP: '/create',
+  SAVED_MAP: '/:id',
```
Comment (VijayanB) [inline maps_dashboards/common/index.ts:23]: nit: EDIT_MAP

File: maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx [hunk 2]
```diff
@@ -63,13 +87,21 @@ const LayerControlPanel = ({ maplibreRef }: Props) => {
...
 
-  const [layers, setLayers] = useState<ILayerConfig[]>([initialLoadLayer]);
-
+  // Initially load the layers from the saved object
   useEffect(() => {
     maplibreRef.current?.on('load', function () {
-      layers.forEach((layer) => {
-        layersFunctionMap[layer.type]?.initial(maplibreRef, layer);
-      });
+      if (!!mapIdFromUrl) {
```
Comment (VijayanB) [inline maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx:93]: Just confirming, is !! expected?

Comment (junqiu-lei) [reply maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx:93]: `!!` is used to convert non-Boolean type values to Boolean type

File: maps_dashboards/common/index.ts (hunk 1 above)
Comment (junqiu-lei) [reply maps_dashboards/common/index.ts]: Updated.

File: maps_dashboards/public/components/map_top_nav/top_nav_menu.tsx [hunk 3]
```diff
@@ -3,24 +3,66 @@
...
+  const [title, setTitle] = useState<string>('');
+  const [description, setDescription] = useState<string>('');
+
+  const fetchMapObject = useCallback(async (): Promise<
+    SimpleSavedObject<MapSavedObjectAttributes>
+  > => {
+    return await savedObjectsClient.get<MapSavedObjectAttributes>('map', mapIdFromUrl);
+  }, [savedObjectsClient, mapIdFromUrl]);
+
+  useEffect(() => {
```
Comment (ashwin-pc) [inline maps_dashboards/public/components/map_top_nav/top_nav_menu.tsx:41]: Again, you should not have to fetch saved object data in each component

File: maps_dashboards/public/components/app.tsx [hunk 4]
```diff
@@ -22,7 +22,7 @@ export const MapsDashboardsApp = () => {
       <I18nProvider>
         <div>
           <Switch>
-            <Route path={APP_PATH.CREATE_MAP} render={() => <MapPage />} />
+            <Route path={[APP_PATH.CREATE_MAP, APP_PATH.SAVED_MAP]} render={() => <MapPage />} />
             <Route exact path="/" render={() => <MapsList />} />
```
[the remark this answers is not in the document]
Comment (junqiu-lei) [reply maps_dashboards/public/components/app.tsx:26]: Will update!

File: maps_dashboards/public/components/map_container/map_container.tsx [hunk 5]
```diff
@@ -9,12 +9,19 @@ import { Map as Maplibre, NavigationControl } from 'maplibre-gl';
 import { LayerControlPanel } from '../layer_control_panel';
 import './map_container.scss';
 import { MAP_INITIAL_STATE, MAP_GLYPHS, MAP_VECTOR_TILE_DATA_SOURCE } from '../../../common';
+import { ILayerConfig } from '../../model/ILayerConfig';
 
-export const MapContainer = () => {
+interface MapContainerProps {
+  mapIdFromUrl: string;
+  setLayers: (layers: ILayerConfig[]) => void;
+  layers: ILayerConfig[];
```
[the remark this answers is not in the document]
Comment (junqiu-lei) [reply maps_dashboards/public/components/map_container/map_container.tsx:17]: Yes, I was planing to use `useReducer` for concisely pass.

**Comment by junqiu-lei** [comment]
@ashwin-pc Thanks your helpful feedback here and offline discussion!  I updated the PR to make fetch saved object at top level component, and for the layers hook, I found most of components need use it and it doesn't have too much levels of component, I prefer to remain the existing hooks to use.

**Review by VijayanB — APPROVED** [review]

File: maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx [hunk 6]
```diff
@@ -34,46 +37,49 @@ interface MaplibreRef {
 
 interface Props {
   maplibreRef: MaplibreRef;
+  mapIdFromUrl: string;
+  setLayers: (layers: ILayerConfig[]) => void;
+  layers: ILayerConfig[];
 }
 
-const LayerControlPanel = ({ maplibreRef }: Props) => {
+const LayerControlPanel = memo(({ maplibreRef, mapIdFromUrl, setLayers, layers }: Props) => {
```
Comment (ashwin-pc) [inline maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx:45]: Whats the intention behind memoizing this component? Did you observe a performance issue?

File: maps_dashboards/public/model/OSMLayerFunctions.ts [hunk 7]
```diff
@@ -6,118 +6,89 @@ interface MaplibreRef {
...
-            if (mbLayer.type === 'symbol') {
-              return;
-            }
-            maplibreInstance.setPaintProperty(
-              mbLayer.id,
-              `${mbLayer.type}-opacity`,
-              layerConfig.opacity
-            );
-          }
+  initial: async (maplibreRef: MaplibreRef, layerConfig: ILayerConfig) => {
```
Comment (ashwin-pc) [inline maps_dashboards/public/model/OSMLayerFunctions.ts:47]: nit: Would it be better to call this `initialize`?

File: maps_dashboards/public/model/OSMLayerFunctions.ts [hunk 8]
```diff
@@ -6,118 +6,89 @@ interface MaplibreRef {
...
+    if (maplibreRef.current) {
+      fetchStyleLayers().then((styleLayers: LayerSpecification[]) => {
```
Comment (ashwin-pc) [inline maps_dashboards/public/model/OSMLayerFunctions.ts:49]: Nice! much more readable!

File: maps_dashboards/tsconfig.json [hunk 9]
```diff
@@ -2,7 +2,9 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "outDir": "./target",
-    "skipLibCheck": true
+    "skipLibCheck": true,
+    // adding global typings
+    "noImplicitAny": false,
```
Comment (ashwin-pc) [inline maps_dashboards/tsconfig.json:7]: This is not good practice, why is this necessary?

File: maps_dashboards/public/components/map_top_nav/get_top_nav_config.tsx [hunk 10]
```diff
@@ -29,19 +40,36 @@ export const getTopNavConfig = (services: MapServices) => {
...
-            title: onSaveProps.newTitle,
-            description: onSaveProps.newDescription,
-            // TODO: Integrate other attributes to saved object
-          });
-          const id = savedMap.id;
+        const onModalSave = async ({ newTitle, newDescription }: OnSaveProps) => {
+          let newlySavedMap;
+          if (mapIdFromUrl) {
+            // edit existing map
+            newlySavedMap = await savedObjectsClient.update('map', mapIdFromUrl, {
```
Comment (ashwin-pc) [inline maps_dashboards/public/components/map_top_nav/get_top_nav_config.tsx:47]: instead of using the saved object client directly, consider using the `SavedObjectLoader`. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2.4/src/plugins/saved_objects/public/saved_object/saved_object_loader.ts#L36-L37

This lets you use a single saved object loader instace across the app that always operates on just the `map` saved object type. It is also Typesafe and you can avoid such if else conditions.

https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/061e5bd8706c3a6a1aa089e6490c56397aa6b60a/src/plugins/vis_builder/public/saved_visualizations/_saved_vis.ts#L14

This can also e done in a future iteration but its one thing that will make working with saved objects a lot more straightforward, especially when it comes to future integrations with the SaveObjectManagement plugin, etc.

File: maps_dashboards/public/components/map_page/map_page.tsx [hunk 11]
```diff
@@ -3,15 +3,46 @@
...
+import { SimpleSavedObject } from 'opensearch-dashboards/public';
 import { MapContainer } from '../map_container';
 import { MapTopNavMenu } from '../map_top_nav';
+import { ILayerConfig } from '../../model/ILayerConfig';
+import { MapServices } from '../../types';
+import { useOpenSearchDashboards } from '../../../../../src/plugins/opensearch_dashboards_react/public';
+import { MapSavedObjectAttributes } from '../../../common/map_saved_object_attributes';
 
 export const MapPage = () => {
+  const [layers, setLayers] = useState<ILayerConfig[]>([]);
```
[the remark this answers is not in the document]
Comment (ashwin-pc) [reply maps_dashboards/public/components/map_page/map_page.tsx:17]: I see you hav made the change to only load the saved object data once, nice! Not a blocker, but this is still prop drilling and not using react context. And the component consuming the layers is still 2 levels deep.

File: maps_dashboards/public/components/map_container/map_container.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashwin-pc) [reply maps_dashboards/public/components/map_container/map_container.tsx:17]: I wouldnt mark this comment as resolved since you mention here that you want to use useReducer, but that change has not been made yet.

**Review by ashwin-pc — COMMENTED** [review]
The updates look nice! The code is much more readable. A few additional comments, some suggestions and a few questions.

File: maps_dashboards/tsconfig.json (hunk 9 above)
Comment (junqiu-lei) [reply maps_dashboards/tsconfig.json:7]: will remove it

Comment (junqiu-lei) [reply maps_dashboards/tsconfig.json]: removed

File: maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx (hunk 6 above)
Comment (ashwin-pc) [reply maps_dashboards/public/components/layer_control_panel/layer_control_panel.tsx:45]: Makes sense :) thanks

File: maps_dashboards/tsconfig.json (hunk 9 above)
Comment (ashwin-pc) [reply maps_dashboards/tsconfig.json]: Oh man, thanks for the callout. should remove it there as well. Thanks.

File: maps_dashboards/public/components/map_top_nav/get_top_nav_config.tsx (hunk 10 above)
Comment (ashwin-pc) [reply maps_dashboards/public/components/map_top_nav/get_top_nav_config.tsx:47]: Yeah that confused me a little. We dont have the best guidance on saved objects right now so thats on us to improve. The loader makes sense in your case since you always interact with a single saved object type to load and save all map data, a use case best suited for the loader. Backlog is good for now.

**Review by ashwin-pc — APPROVED** [review]
LGTM

**Review by VijayanB — APPROVED** [review]
